### PR TITLE
Add support for cracking files using GPG symmetric encryption

### PR DESCRIPTION
**Pull request by kholia**
@magnumripper I need help in locating, and linking against the GPGME library.

This pull request includes "gpgme.m4", which might be of some help.

### CODE REVIEW

**Comment by magnumripper** [comment]
What's the actual encryption involved? Do we really need to include an external lib?

**Comment by kholia** [comment]
```
$ pgpdump password.asc 
Old: Symmetric-Key Encrypted Session Key Packet(tag 3)(13 bytes)
	New version(4)
	Sym alg - AES with 256-bit key(sym 9)
	Iterated and salted string-to-key(s2k 3):
		Hash alg - SHA1(hash 2)
		Salt - 20 53 8c 8d 69 96 4d 96 
		Count - 65536(coded count 96)
New: Symmetrically Encrypted and MDC Packet(tag 18)(64 bytes)
	Ver 1
	Encrypted data [sym alg is specified in sym-key encrypted session key]
		(plain text + MDC SHA1(20 bytes))
```

We might just be able to crack this stuff without using an external library. Thanks! :)

**Comment by kholia** [comment]
@magnumripper this should be working now, please review! :)

**Comment by kholia** [comment]
https://github.com/kholia/PGPCrack-NG/tree/master/testdata has some samples files (all of them aren't supported currently).

**Comment by magnumripper** [comment]
Hmm those bot failures are unrelated and my fault, but I never got any mail about them. Fixed now.

**Comment by magnumripper** [comment]
We should only use external code when we really really have to!

File: src/gpg_fmt_plug.c [hunk 1]
```diff
@@ -402,29 +413,37 @@ static int valid(char *ciphertext, struct fmt_main *self)
...
 			goto err;
-		p = strtokm(NULL, "*");  /* NOTE, do not goto err if null, we WANT p nul if there are no fields */
 	}
 
-	if (p)	/* at this point, there should be NO trailing stuff left from the hash. */
-		goto err;
+	if (symmetric_mode) {
+		// fprintf(stderr, "GPG symmetric mode support is in progress!\n");
+		// goto err;
+	}
```
Comment (magnumripper) [inline src/gpg_fmt_plug.c]: You need to clean up all these weird artifacts before merging. This is an empty body. Also, lots of commented-out stuff should actually be removed once it's working.

**Comment by magnumripper** [comment]
This is cool! It looks fine except you need to clean it up from artifacts.

**Comment by kholia** [comment]
@magnumripper hopefully, the current version looks better.

**Comment by kholia** [comment]
Seems to work fine on real challenges too  :)

```
$ ../run/john --rules -w=all hash # https://id0-rsa.pub/problem/1/
Using default input encoding: UTF-8
Loaded 1 password hash (gpg, OpenPGP / GnuPG Secret Key [32/64])
Will run 2 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:06 0,00% (ETA: 2015-11-17 09:26) 0g/s 2177p/s ...
seaXXXXXXX       (?)
```

**Comment by magnumripper** [comment]
> seaXXXXXXX       (?)

Awesome. Unless it's a false positive? Run in for a while with `--keep-running` and verify you don't get more candidates.

**Comment by magnumripper** [comment]
*The password used was a single English word. Recover the message.*

seaXXXXXXX smells false positive to me. I hope I'm wrong!

**Comment by kholia** [comment]
I masked the actual password (didn't want to spoil the challenge) ;)

The actual password, and the decrypted message is fine. We have a 20-byte
SHA-1 verifier in this file.

**Comment by magnumripper** [comment]
Awsome. And I also don't see any false positives.

Speed is very low, what's the KDF here?

**Comment by magnumripper** [comment]
Ah, yes I cracked the real one now.

**Comment by kholia** [comment]
@magnumripper the KDF is "S2KItSaltedSHA1Generator" with count = 65536 (I can't recall what the count actually meant), can we do some SIMD magic here? :-)

**Comment by magnumripper** [comment]
Yeah, I saw that. It's not actually 64K calls to the digest, but 64K updates. For a length-8 password, it's about 8K calls to actual SHA-1 digest. So speed is pretty good.

But yes, we could definitely do 4x or 8x by using SIMD.

**Pull request merged**